### ISSUE TITLE
Start options to set the location from individual files xdg-data-home files + profile support

### ISSUE THREAD

**Issue by Ambrevar** [atlas-engineer/next#646]
Currently we have only two knobs to move the xdg-data-home files (sessions, cookies, hostlist, history, bookmarks, socket).  XDG_DATA_HOME and --session.

The downside it that it's tricky to use a different file for only one component, say, use a different cookie file but the same history.

We already have configurable knobs for sessions and the init file (which is not in XDG_DATA_HOME, but the mechanism is the same).

1. Let's implement a general way to configure the location of all files individually.

2. It'd be nice if the `start` function would allow setting the same knobs as the command line, so that we have the same flexibility from the REPL.  In particular, it's useful to set the socket file from the REPL to avoid calling the binary single instance that could be running.

**Comment by jmercouris** [comment]
1. Is too abstract for me to understand, can you please elaborate on how you imagine that may look?

2. OK!

**Comment by Ambrevar** [comment]
Actually I've fixed 2. yesterday! :)

About 1, the goals are:

- Allow precise and configuration of where the individual persisted data
  is stored.

- Make it controllable from the command line.  The point is that the
  user should not have to write Lisp code just to change the path of the
  persisted data files.

Suggested strategy to avoid a profusion of command line options:

- Allow configuring the "next" part of "$XDG_DATA_HOME/next".

- Use the same session path expansion function for bookmarks, cookies, etc.

- Remove the --session option.

- Add a --profile command line option.

- Add a PROFILE slot for the BROWSER class and move bookmarks-path,
  cookies-path, etc. in there.
  This slot would be an association between a profile name and paths for
  bookmarks, cookies, etc.

When starting Next, the browser is instantiated with the profile
matching "--profile $PROFILE_NAME", or the "default" profile otherwise.

In their init file, users could define their own profiles, i.e.

```
'(("default"  ...)
  ("private" nil) ; no paths, no data is persisted.
  ("work" (:bookmarks-path "work-bookmarks" :cookie-path "..." ...)))
```

It's important to get this right before 2.0 cause it would be
API-breaking afterwards.

Another cool feature would be to add a command to start Next with a
given profile.  Many browsers can "open a private window", this would
allow us to do this (and more).

Thoughts?

**Comment by Ambrevar** [comment]
Note: we also need to add a --socket / --no-socket option, since it's
the only file with the init file that's required before the browser is
instantiated.

--no-socket would allow to start Next in multi-instance.

**Comment by Ambrevar** [comment]
Maybe add a `--list-profiles` option?

Could be nice to add a documentation string to profile so that we can explain what they do from the command line.

Should we make a profile class?

**Comment by jmercouris** [comment]
A profiles class would seem prudent for some of the common location configurations to encapsulate them in a neat bundle.

The name profile suggests to the user also a set of settings associated with a persona. Thusly I suggest data-root or something like that.

I also think that it could be cool to allow for this to be prompted interactively as we now do with restoring buffers. IE, use a profile? Which one?

**Comment by jmercouris** [comment]
The reason why I point out the necessity of not naming it profiles is because I believe we have a very powerful usecase with profiles as actual sets of behaviors and locations:


example a define-configuration profile for work, or for gaming, with different settings and toggles

**Comment by Ambrevar** [comment]
> A profiles class would seem prudent for some of the common location configurations to encapsulate them in a neat bundle.

Yes!

> The name profile suggests to the user also a set of settings associated with a persona. Thusly I suggest data-root or something like that.

Agreed, I was thinking the same thing.  Not sure `data-root` would tell
much to the newcomer.  `data-storage`?

> I also think that it could be cool to allow for this to be prompted interactively as we now do with restoring buffers. IE, use a profile? Which one?

Why not!

**Comment by jmercouris** [comment]
Data-storage seems a fine name :-)

**Comment by Ambrevar** [comment]
But `--list-data-storages` does not roll off the tongue :p
What about `--data-location` and `--list-data-locations` then?

**Comment by jmercouris** [comment]
`data-path` and `list-data-paths` et voila!

**Comment by Ambrevar** [comment]
nitpicking, but "data-path" may mislead the user into thinking that the option takes
a path argument, while it could also be an entry in a slot of the
BROWSER class that refers to a collection of paths.

If you don't have a better suggestion, let's go with data-path and make
it clear in `next --help` that both paths and well-known names are supported.

**Comment by jmercouris** [comment]
That’s the best I have for now, we can rename it in time, query-replace it away :-)

**Comment by Ambrevar** [comment]
I've been thinking of an implementation.

Goals:
- Paths must be customizable from the init.lisp.
- Paths must be customizable from the command line.
- Paths must be customizable by "profiles", i.e. a well known name like "default", "private", "work" that groups settings together.
- Third-party extensions must be able to expose their customizable paths to the user so that it still composes with profiles.

We have 2 concepts: profiles and data-paths.
Their configuration must compose, so we must be able to customize a data-path for any profile (or regardless of any profile) and vice versa.

Common Lisp gives us the tool for this composition: multi-dispatch.  So we would have a macro

```lisp
(defmacro with-data-path (datapath)
 `(progn
   ...
    (open-data-path (current-profile) datapath)
    ...))
```

and `open-data-path` would be a method on the `data-profile` class and `data-path` class.  This way, it's possible to specialize the method for any profile / data-path.= combination.

By default,  we would provide two methods, one for the "default" profile, one for the "private" profile, both working for any data path.

The default implementation would do something like `(or (getf *options* :data-path) data-path)` so that we can use a path given on the command line.
Because the command line is not customizable from the init.lisp, it's not possible to add new data-paths customization arguments.  So instead I propose that set the data path from a name, life this:

```sh
next --with-data-path bookmarks=/my/path --with-data-path cookies=/my/other/path
```

Note: this is something Emacs didn't get right. As a result, it's cumbersome to configure where files are persisted in Emacs.  It'd be great if Next would get this right! :)

**Comment by Ambrevar** [comment]
Merged with e5cedaf9de1181825e9ec1362af1a3fd42586f44.
